pkg/asset/installconfig/alibabacloud/validation.go

```diff
@@ -16,76 +15,107 @@ func Validate(client *Client, ic *types.InstallConfig) error {
+	allErrs := field.ErrorList{}
+	var mpool mergedMachinePool
 
 	for idx, compute := range ic.Compute {
-		fldPath := field.NewPath("compute").Index(idx)
+		computePoolFieldPath := field.NewPath("compute").Index(idx).Child("platform", "alibabacloud")
+		defaultPool := alibabacloudtypes.DefaultWorkerMachinePoolPlatform()
+		mpool.setWithFieldPath(&defaultPool, computePoolFieldPath)
+		if ic.Platform.AlibabaCloud != nil {
+			mpool.setWithFieldPath(ic.Platform.AlibabaCloud.DefaultMachinePlatform, field.NewPath("platform", "alibabacloud", "defaultMachinePlatform"))
+		}
+
 		if compute.Platform.AlibabaCloud != nil {
-			allErrs = append(allErrs, validateMachinePool(client, ic, fldPath.Child("platform", "alibabacloud"), compute.Platform.AlibabaCloud, compute.Replicas)...)
+			mpool.setWithFieldPath(compute.Platform.AlibabaCloud, computePoolFieldPath)
 		}
+		allErrs = append(allErrs, validateMachinePool(client, ic, &mpool)...)
 	}
-
-	return allErrs.ToAggregate()
+	return allErrs
 }
 
-func validatePlatform(client *Client, ic *types.InstallConfig, path *field.Path) field.ErrorList {
+func validateMachinePool(client *Client, ic *types.InstallConfig, pool *mergedMachinePool) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if ic.AlibabaCloud.ResourceGroupID != "" {
-		allErrs = append(allErrs, validateResourceGroup(client, ic, path)...)
-	}
+	if len(pool.Zones) > 0 {
+		availableZones := map[string]bool{}
 
-	if ic.Platform.AlibabaCloud.DefaultMachinePlatform != nil {
-		allErrs = append(allErrs, validateMachinePool(client, ic, path.Child("defaultMachinePlatform"), ic.Platform.AlibabaCloud.DefaultMachinePlatform, nil)...)
+		response, err := client.DescribeAvailableResource("Zone")
+		if err != nil {
+			return append(allErrs, field.InternalError(pool.zonesFieldPath, err))
+		}
+		for _, availableZone := range response.AvailableZones.AvailableZone {
+			if availableZone.Status == "Available" {
+				availableZones[availableZone.ZoneId] = true
+			}
+		}
+
+		for idx, zone := range pool.Zones {
+			if !availableZones[zone] {
+				allErrs = append(allErrs, field.Invalid(pool.zonesFieldPath.Index(idx), zone, fmt.Sprintf("zone ID is unavailable in region %q", ic.Platform.AlibabaCloud.Region)))
+			}
+		}
+	}
+	// InstanceType and zones are related.
+	// If the availability zone is not available, the instanceType will not be validated.
+	if len(allErrs) == 0 {
+		allErrs = append(allErrs, validateInstanceType(client, pool.Zones, pool.InstanceType, pool.instanceTypeFieldPath)...)
 	}
 
 	return allErrs
 }
 
-func validateMachinePool(client *Client, ic *types.InstallConfig, fldPath *field.Path, pool *alibabacloudtypes.MachinePool, replicas *int64) field.ErrorList {
+func validateInstanceType(client *Client, zones []string, instanceType string, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	var zones []string
-	response, err := client.DescribeAvailableResource("Zone")
+	zoneWithStock := map[string]bool{}
```

Comment:
I think I steered you to `map[string]bool` here, but it would be a bit cleaner if you used `sets.NewString` instead.
https://github.com/kubernetes/apimachinery/blob/5f072755815add51938b52f806c555ed8c86a1ce/pkg/util/sets/string.go#L30
```
zonesWithStock := sets.NewString(availableZones)
```
```
for_, zoneID := range zones {
  if zonesWithStock.Has(zoneID) {
    ...
  }
}
```